File: src/EmployeeTasks/Grafos.java
package EmployeeTasks;

import AdministratorTasks.TablaHash;
import Service.Estacion;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.PriorityQueue;

public class Grafos {

    private TablaHash<Estacion, Integer> estaciones;
    private ArrayList<ArrayList<Arista>> listaAdyacencia;

    public Grafos(int numeroEstaciones) {
        this.estaciones = new TablaHash<>(numeroEstaciones);
        this.listaAdyacencia = new ArrayList<>();
    }

    public TablaHash<Estacion, Integer> getEstaciones() {
        return estaciones;
    }

    public void agregarEstacion(Estacion estacion) {
        if (estacion == null) {
            System.out.println("La estación no puede ser nula.");
            return;
        }

        if (!estaciones.contieneClave(estacion)) {
            estaciones.insertarValores(estacion, listaAdyacencia.size());
            listaAdyacencia.add(new ArrayList<>());
            System.out.println("Estación agregada: " + estacion.getNombreEstacion());
        }
        else {
            System.out.println("La estación ya existe: " + estacion.getNombreEstacion());
        }
    }

    public Estacion buscarEstacion(String id) {
        for (Estacion estacion : estaciones.obtenerClaves()) {
            if (estacion.getNombreEstacion().equals(id)) {
                return estacion;
            }
        }
        System.out.println("No se encontró la estación con id: " + id);
        return null;
    }

    public void agregarRutaDirigida(Estacion origen, Estacion destino, int distancia) {
        if (!estaciones.contieneClave(origen) || !estaciones.contieneClave(destino)) {
            System.out.println("Las estaciones de origen o destino no existen.");
            return;
        }

        int indiceOrigen = estaciones.buscarValores(origen);
        int indiceDestino = estaciones.buscarValores(destino);

        if (indiceOrigen >= 0 && indiceOrigen < listaAdyacencia.size()) {
            listaAdyacencia.get(indiceOrigen).add(new Arista(indiceDestino, distancia, 1));
            System.out.println("Ruta dirigida agregada: " + origen.getNombreEstacion() + " -> " + destino.getNombreEstacion() + " | Distancia: " + distancia + " km");
        }
        else {
            System.out.println("Error: Índice de origen fuera de rango.");
        }
    }

    public void agregarRutaNoDirigida(Estacion origen, Estacion destino, int distancia) {
        if (!estaciones.contieneClave(origen) || !estaciones.contieneClave(destino)) {
            System.out.println("Las estaciones de origen o destino no existen.");
            return;
        }

        int indiceOrigen = estaciones.buscarValores(origen);
        int indiceDestino = estaciones.buscarValores(destino);

        if (indiceOrigen >= 0 && indiceOrigen < listaAdyacencia.size() && indiceDestino >= 0 && indiceDestino < listaAdyacencia.size()) {
            listaAdyacencia.get(indiceOrigen).add(new Arista(indiceDestino, distancia, 1));
            listaAdyacencia.get(indiceDestino).add(new Arista(indiceOrigen, distancia, 1));
            System.out.println("Ruta no dirigida agregada: " + origen.getNombreEstacion() + " <-> " + destino.getNombreEstacion() + " | Distancia: " + distancia + " km");
        }
        else {
            System.out.println("Error: Índices fuera de rango.");
        }
    }

    public ArrayList<Integer> aplicarDijkstra(Estacion estacionOrigen) {
        if (!estaciones.contieneClave(estacionOrigen)) {
            System.out.println("La estación de origen no existe.");
            return null;
        }

        int indiceOrigen = estaciones.buscarValores(estacionOrigen);

        ArrayList<Integer> distanciasMinimas = new ArrayList<>();
        ArrayList<Boolean> estacionesVisitadas = new ArrayList<>();

        for (int i = 0; i < listaAdyacencia.size(); i++) {
            distanciasMinimas.add(Integer.MAX_VALUE);
            estacionesVisitadas.add(false);
        }

        PriorityQueue<Arista> colaPrioridad = new PriorityQueue<>();
        colaPrioridad.add(new Arista(indiceOrigen, 0, 0));
        distanciasMinimas.set(indiceOrigen, 0);

        while (!colaPrioridad.isEmpty()) {
            Arista rutaActual = colaPrioridad.poll();
            int estacionActual = rutaActual.getDestino();

            if (estacionesVisitadas.get(estacionActual)) continue;
            estacionesVisitadas.set(estacionActual, true);

            Iterator<Arista> iterador = listaAdyacencia.get(estacionActual).iterator();
            while (iterador.hasNext()) {
                Arista rutaVecina = iterador.next();
                int estacionVecina = rutaVecina.getDestino();
                int distanciaRuta = rutaVecina.getDistancia();

                if (!estacionesVisitadas.get(estacionVecina) && distanciasMinimas.get(estacionActual) + distanciaRuta < distanciasMinimas.get(estacionVecina)) {
                    distanciasMinimas.set(estacionVecina, distanciasMinimas.get(estacionActual) + distanciaRuta);
                    colaPrioridad.add(new Arista(estacionVecina, distanciasMinimas.get(estacionVecina), 0));
                }
            }
        }

        System.out.println("Distancias más cortas desde " + estacionOrigen.getNombreEstacion() + ":");
        for (Estacion estacion : estaciones.obtenerClaves()) {
            int indice = estaciones.buscarValores(estacion);
            if (indice >= 0 && indice < distanciasMinimas.size()) {
                System.out.println("- " + estacion.getNombreEstacion() + ": " + distanciasMinimas.get(indice) + " km");
            }
            else {
                System.out.println("- " + estacion.getNombreEstacion() + ": No se pudo calcular la distancia.");
            }
        }

        return distanciasMinimas;
    }

    public int calcularDistanciaEntre(Estacion origen, Estacion destino) {
        if (!estaciones.contieneClave(origen) || !estaciones.contieneClave(destino)) {
            System.out.println("Error: Una de las estaciones no existe.");
            return -1;
        }

        ArrayList<Integer> distancias = aplicarDijkstra(origen);
        int indiceDestino = estaciones.buscarValores(destino);

        if (indiceDestino >= 0 && indiceDestino < distancias.size()) {
            return distancias.get(indiceDestino);
        }

        else {
            System.out.println("No se pudo calcular la distancia entre " + origen.getNombreEstacion() + " y " + destino.getNombreEstacion());
            return -1;
        }
    }
    
}

File: Códigos/TareasEmpleado/Grafos.cs
using System;
using System.Collections.Generic;
using ProyectoEstructuras.Códigos.Servicio;
using ProyectoEstructuras.Códigos.TareasAdministrador;
using ProyectoEstructuras.Códigos.TareasEmpleado.ProyectoEstructuras.Códigos.TareasEmpleado;

namespace ProyectoEstructuras.Códigos.TareasEmpleado
{
    public class Grafos
    {
        private TablaHash<Estacion, int> estaciones;
        private List<List<Arista>> listaAdyacencia;

        public TablaHash<Estacion, int> Estaciones { get => estaciones; set => estaciones = value; }
        public List<List<Arista>> ListaAdyacencia { get => listaAdyacencia; set => listaAdyacencia = value; }

        public Grafos(int numeroEstaciones)
        {
            this.Estaciones = new TablaHash<Estacion, int>(numeroEstaciones);
            this.ListaAdyacencia = new List<List<Arista>>();

            for (int ii = 0; ii < numeroEstaciones; ii++)
            {
                ListaAdyacencia.Add(new List<Arista>());
            }
        }

        public void AgregarEstacion(Estacion estacion)
        {
            if (estacion == null)
            {
                return;
            }

            if (!Estaciones.ContieneClave(estacion))
            {
                Estaciones.InsertarValores(estacion, ListaAdyacencia.Count);
                ListaAdyacencia.Add(new List<Arista>());
            }
        }

        public Estacion BuscarEstacion(string id)
        {
            foreach (Estacion estacion in Estaciones.ObtenerClaves())
            {
                if (estacion.NombreEstacion.Equals(id))
                {
                    return estacion;
                }
            }
            return null;
        }

        public void AgregarRutaDirigida(Estacion origen, Estacion destino, int distancia)
        {
            if (!Estaciones.ContieneClave(origen) || !Estaciones.ContieneClave(destino))
            {
                return;
            }

            int indiceOrigen = Estaciones.BuscarValores(origen);
            int indiceDestino = Estaciones.BuscarValores(destino);

            if (indiceOrigen >= 0 && indiceOrigen < ListaAdyacencia.Count)
            {
                ListaAdyacencia[indiceOrigen].Add(new Arista(indiceDestino, distancia, 1));
            }
        }

        public void AgregarRutaNoDirigida(Estacion origen, Estacion destino, int distancia)
        {
            if (!Estaciones.ContieneClave(origen) || !Estaciones.ContieneClave(destino))
            {
                return;
            }

            int indiceOrigen = Estaciones.BuscarValores(origen);
            int indiceDestino = Estaciones.BuscarValores(destino);

            if (indiceOrigen >= 0 && indiceOrigen < ListaAdyacencia.Count && indiceDestino >= 0 && indiceDestino < ListaAdyacencia.Count)
            {
                ListaAdyacencia[indiceOrigen].Add(new Arista(indiceDestino, distancia, 1));
                ListaAdyacencia[indiceDestino].Add(new Arista(indiceOrigen, distancia, 1));
            }
        }

        public List<int> AplicarDijkstra(Estacion estacionOrigen)
        {
            if (!Estaciones.ContieneClave(estacionOrigen))
            {
                return null;
            }

            int indiceOrigen = Estaciones.BuscarValores(estacionOrigen);

            List<int> distanciasMinimas = new List<int>(new int[ListaAdyacencia.Count]);
            List<bool> estacionesVisitadas = new List<bool>(new bool[ListaAdyacencia.Count]);

            for (int i = 0; i < ListaAdyacencia.Count; i++)
            {
                distanciasMinimas[i] = int.MaxValue;
            }

            PriorityQueue<Arista> colaPrioridad = new PriorityQueue<Arista>();
            colaPrioridad.Enqueue(new Arista(indiceOrigen, 0, 0), 0);
            distanciasMinimas[indiceOrigen] = 0;

            while (colaPrioridad.Count > 0)
            {
                Arista rutaActual = colaPrioridad.Dequeue();
                int estacionActual = rutaActual.Destino;

                if (estacionesVisitadas[estacionActual]) continue;

                estacionesVisitadas[estacionActual] = true;

                foreach (Arista rutaVecina in ListaAdyacencia[estacionActual])
                {
                    int estacionVecina = rutaVecina.Destino;
                    int distanciaRuta = rutaVecina.Distancia;

                    if (!estacionesVisitadas[estacionVecina] &&
                        distanciasMinimas[estacionActual] + distanciaRuta < distanciasMinimas[estacionVecina])
                    {
                        distanciasMinimas[estacionVecina] = distanciasMinimas[estacionActual] + distanciaRuta;
                        colaPrioridad.Enqueue(new Arista(estacionVecina, distanciasMinimas[estacionVecina], 0), distanciasMinimas[estacionVecina]);
                    }
                }
            }

            Console.WriteLine($"Distancias más cortas desde {estacionOrigen.NombreEstacion}:");
            foreach (var estacion in estaciones.ObtenerClaves())
            {
                int indice = estaciones.BuscarValores(estacion);
                if (indice >= 0 && indice < distanciasMinimas.Count)
                {
                    Console.WriteLine($"- {estacion.NombreEstacion}: {distanciasMinimas[indice]} km");
                }
                else
                {
                    Console.WriteLine($"- {estacion.NombreEstacion}: No se pudo calcular la distancia.");
                }
            }

            return distanciasMinimas;
        }

        public int CalcularDistanciaEntre(Estacion origen, Estacion destino)
        {
            if (!Estaciones.ContieneClave(origen) || !Estaciones.ContieneClave(destino))
            {
                return -1;
            }

            List<int> distancias = AplicarDijkstra(origen);
            int indiceDestino = Estaciones.BuscarValores(destino);

            if (indiceDestino >= 0 && indiceDestino < distancias.Count)
            {
                return distancias[indiceDestino];
            }
            return -1;
        }
    }
}